core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -612,6 +615,9 @@ protected StructReader(
         } else if (field.fieldId() == MetadataColumns.IS_DELETED.fieldId()) {
           positionList.add(pos);
           constantList.add(false);
+        } else if (record.getField(field.name()) == null && field.initialDefault() != null) {
```

Comment:
I think the reason for leaving out the field in `BuildAvroProjection` was to make this check work. But I'm pretty sure that the Avro Schema needs to match the Iceberg Schema. So this will probably need to have some other way of detecting that the field will not be read from the field because it was missing from the file schema.

Reply:
Sure, I suggested the above way of creating a "_d" field which can be a marker, or I can use a custom avro property, something like "<HAS_DEFAULT -> true>" to signify it, WDYT?

Reply:
Passing information through field names is error prone because you don't control the names of real columns.

Instead, what about adding a metadata key to the Avro schema field that signals this?

Reply:
In the latest patch, `record` is no longer required in this method. The default value is set here similar using a mechanism similar to `positions` and `constants` and then it is used in the `read()` method. Removing dependency on `record` simplified this API and a few others as `Schema record` is no longer required in the signature (similar to before this patch).